verify-logging/main.go

```diff
@@ -0,0 +1,200 @@
+	flag.StringVar(&metricType, "metric-type", "", "The [metric type](https://cloud.google.com/monitoring/mql/reference#metric-tabop) to get from the table-name")
+	flag.StringVar(&predicates, "predicates", "", "Commma delimited list of [predicates](https://cloud.google.com/monitoring/mql/reference#filter-tabop)")
+	flag.StringVar(&responseCodeClass, "response-code-class", "5xx", "The response_code_class that is being monitored for the error condition")
+	flag.Float64Var(&maxErrorPercentage, "max-error-percentage", 10, "The maximum allowable percentage of the specified response_code_class per sliding window")
+	flag.DurationVar(&slidingWindow, "sliding-window", time.Minute, "The duration of the sliding window")
+	flag.DurationVar(&triggerDuration, "trigger-duration", 5*time.Minute, "The time required to observe the error condition for verify to fail")
+	flag.DurationVar(&timeToMonitor, "time-to-monitor", 20*time.Minute, "The time to monitor for response failures before the verification is marked successful")
+	flag.DurationVar(&refreshPeriod, "refresh-period", 5*time.Minute, "The time to wait before refreshing the data set with new data")
+	flag.StringVar(&customQuery, "custom-query", "", "Customized query following [MQL](https://cloud.google.com/monitoring/mql/reference) to use for query instead. By specifying this, the query will not be crafted by the program")
+
+	flag.Parse()
+
+	fmt.Println("---")
+	fmt.Println("Verification configured as follows:")
+	fmt.Printf("Project: %q\n", project)
+	fmt.Println(formatMsg(fmt.Sprintf("Table Name: %q", tableName)))
+	fmt.Println(formatMsg(fmt.Sprintf("Metric Type: %q", metricType)))
+	fmt.Println(formatMsg(fmt.Sprintf("Predicates: %q", predicates)))
+	fmt.Println(formatMsg(fmt.Sprintf("Response Code Class: %q", responseCodeClass)))
+	fmt.Printf("Max Error Percentage: %v\n", maxErrorPercentage)
+	fmt.Println(formatMsg(fmt.Sprintf("Sliding Window: %v", slidingWindow)))
+	fmt.Printf("Trigger Duration: %v\n", triggerDuration)
+	fmt.Printf("Time To Monitor: %v\n", timeToMonitor)
+	fmt.Printf("Refresh Period: %v\n", refreshPeriod)
+	fmt.Println("---")
+}
+
+func main() {
+	ctx := context.Background()
+
+	client, err := monitoring.NewQueryClient(ctx)
+	if err != nil {
+		fmt.Printf("Unable to create NewQueryClient: %v\n", err)
+		os.Exit(1)
+	}
+	defer client.Close()
+
+	timeToStart := time.Now()
+	timeToEnd := timeToStart.Add(timeToMonitor)
+
+	queryToUse := getQueryText(timeToStart)
+	fmt.Printf("The query is %q\n", queryToUse)
+
+	refreshCount := 1
+	for time.Now().Before(timeToEnd) {
+		triggered, err := errorConditionTriggered(ctx, client, refreshCount, queryToUse)
+		if err != nil {
+			fmt.Printf("Failed to determine whether error condition triggered %v\n", err)
+			os.Exit(1)
+		}
+		if triggered {
+			fmt.Printf("Failing verify\n")
+			os.Exit(1)
+		}
+		time.Sleep(refreshPeriod)
+		refreshCount++
+	}
+	fmt.Printf("Done\n")
+	os.Exit(0)
```

Comment:
Is there a reason we need to call os.exit here like this? It has some side effects that I don't think matter here (ex skipping deferred functions) and its not clear if we are trying to cause any of them

Reply:
Understood, factoring out into helper func to allow for deferred execution even if error is thrown.